Improve release process instruction (#261)
kyma-project/community: guidelines/releases/release-process.md

## Preparation

This section only applies to new major and minor versions. If you do a patch release, skip this paragraph.

Preparation consists of defining new release jobs and removing old ones.

### Define release jobs

Define the release jobs on the `master` branch in the `test-infra` repository. Since every job name has to be unique, prefix it with `pre-rel{XY}`. To learn how to define a release job for a component, read the following [document](https://github.com/kyma-project/test-infra/blob/master/docs/prow/release-jobs.md). To find all release jobs for the 0.9 release, look for job names with the `pre-rel09` prefix.

1. In `test-infra` repository navigate to `prow/jobs` directory.

> 1. Define release jobs in the `test-infra` repository in the following files:

Author: What about now? 🙂

   - `test-infra/watch-pods.yaml`
   - every YAML in `kyma/components` 
   - every YAML in `kyma/tests`
   - every YAML in `kyma/tools/` **except**:
     - `prow/jobs/kyma/tools/docsbuilder/docsbuilder.yaml`
     - `prow/jobs/kyma/tools/failery/failery.yaml`
   - `kyma/kyma-docs.yaml`
   - `kyma/kyma-integration.yaml`
   - `kyma/kyma-gke-integration.yaml`
   - `kyma/kyma-gke-upgrade.yaml`
   - `kyma/kyma-artifacts.yaml`
   - `kyma/kyma-installer.yaml`
   - `kyma/kyma-github-release.yaml`

>         > **Note:** Remember about modifying `presets` array. For example, for 0.9 release the preset is the following:

Reviewer: ```suggestion
        > **Note:** Remember to modify the `presets` array. This is an example of the 0.9 release `preset`:
```

        ```
        - labels:
            preset-target-commit-0.9: "true"
            env:
            - name: RELEASE_TARGET_COMMIT
                value: release-0.9
        ```

> 1. Ensure that tests for release jobs exist. Release tests usually iterate through all release versions and run tests for them.

Reviewer: ```suggestion
1. Ensure that tests for the release jobs exist. Release tests usually iterate through all release versions and run tests for them.
```

See the `TestBucReleases` test defined in `development/tools/jobs/kyma/binding_usage_controller_test.go` as a reference.

1. Update the `GetAllKymaReleaseBranches()` function
defined in the `development/tools/jobs/tester/tester.go` file under the `test-infra` repository.

1. Define branch protection rules for the release branch in the `prow/config.yaml` file.

    For example, see the release-0.9 definition:

    ```yaml
    release-0.9:
      protect: true
      required_status_checks:
        contexts:
          - pre-rel09-kyma-integration
          - pre-rel09-kyma-gke-integration
          - pre-rel09-kyma-artifacts
          - pre-rel09-kyma-installer
          - pre-rel09-kyma-gke-upgrade
          - pre-rel09-kyma-gke-central-connector
    ```

### Remove previous release jobs

> After adding new release jobs, remove all previous ones. For example, while doing 0.9 release, add `pre-rel09` jobs and remove all `pre-rel06` jobs. Make sure that there are only defined jobs with `pre-rel07`, `pre-rel08` and `pre-rel09` prefixes.

Reviewer: Hmm, so you don't remove ALL the previous jobs as such but you leave those from three releases back, right?

Author: Yes, exactly

Reviewer: ```suggestion
After adding new release jobs, remove the old ones. Remember to leave jobs for three latest releases. For example, during the preparation for the 0.9 release, add `pre-rel09` jobs and remove all `pre-rel06` jobs. Make sure that the only defined jobs are those with `pre-rel07`, `pre-rel08`, and `pre-rel09` prefixes.
```

## Do release

> **NOTE:** Currently, you are most likely required to push to the `release-x.y` branch more than once. Make sure that a user with **admin** role is present in the `kyma` repository. 

> Before doing the release, make sure there is no mismatch between source code and YAML files. All components are rebuilt from source code, which means all Helm charts need to be up to date.

Reviewer: ```suggestion
Before a release, make sure there is no mismatch between source code and `.yaml` files. All components are rebuilt from source code, which requires all Helm charts to be up to date.
```

2. Ensure that the `prow/RELEASE_VERSION` file from the `test-infra` repository on a release branch contains the correct version to be created. The file should contain a release version following the `{A}.{B}.{C}` or `{A}.{B}.{C}-rc{D}` format, where `A`,`B`, `C`, and `D` are numbers. If you define a release candidate version, a pre-release is created.  

    Make sure the `RELEASE_VERSION` file includes just this single line:  
    ```bash
    echo -n {RELEASE_VERSION} > prow/RELEASE_VERSION
    ```

3. Push the branch to the `test-infra` repository.

4. Create a PR to `test-infra/release-x.y`. This triggers the pre-release job for `watch-pods`. 

>     > **NOTE:** If you want to just trigger `watch-pods` build without introducing any changes, edit any file within `test-infra` repository and create a pull request. You don't need to merge it - the job will be triggered anyway. After successful `watch-pods` image build, you can close the pull request.

Reviewer: ```suggestion
    > **NOTE:** To trigger the `watch-pods` build without introducing any changes, edit any file within the `test-infra` repository and create a pull request. You don't need to merge it as the job will be triggered anyway. After a successful `watch-pods` image build, you can close the pull request.
```

> 5. Prepare and merge pull request to `master` with updating RELEASE_VERSION for next minor release candidate. For example, if `RELEASE_VERSION` on `master` branch is set to `0.9.2`, then change the version to `1.0.0-rc1`.

Reviewer: What do you mean by preparing the pull request? Is the `RELEASE_VERSION` a file? Would you agree with the below?
```suggestion
5. Update the `RELEASE_VERSION`file with the name of the next minor release candidate and merge the pull request to `master`. For example, if the `RELEASE_VERSION` on the `master` branch is set to `0.9.2`, then change the version to `1.0.0-rc1`.
```

### kyma-project/kyma

1. Create a release branch in the `kyma` repository.

    Do it only for major and minor releases.
    The name of this branch should follow the `release-x.y` pattern, such as `release-0.9`.

    ```bash
    git fetch upstream
    git checkout -b release-{release_version} upstream/master
    git push upstream release-{release_version}
    ```

>     > **NOTE:** Pushing anything to release branch (and also pushing rebased branch) results in creating new Github release.

Reviewer: I'm not sure what you meant, could you rephrase?
```suggestion
    > **NOTE:** If you push anything to the release branch, including the rebased branch, a new Github release is created.
```

2. Ensure that the `prow/RELEASE_VERSION` file from the `test-infra` repository on a release branch contains the correct version to be created. The file should contain a release version following the `{A}.{B}.{C}` or `{A}.{B}.{C}-rc{D}` format, where `A`,`B`, `C`, and `D` are numbers. If you define a release candidate version, a pre-release is created.  

    Make sure the `RELEASE_VERSION` file includes just this single line:  
    ```bash
    echo -n {RELEASE_VERSION} > prow/RELEASE_VERSION
    ```

3. Push the branch to the `test-infra` repository.

4. Create a PR to `test-infra/release-x.y`. This triggers the pre-release job for `watch-pods`. 

>     > **NOTE:** To trigger the `watch-pods` build without introducing any changes, edit any file within the `test-infra` repository and create a pull request. You don't need to merge it as the job will be triggered anyway. After a successful `watch-pods` image build, you can close the pull request.

Reviewer: ```suggestion
    > **NOTE:** To trigger the `watch-pods` build without introducing any changes, edit any file within the `test-infra` repository and create a pull request. You don't need to merge it as the job is triggered anyway. After a successful `watch-pods` image build, close the pull request.
```

        ```yaml
        dir:
        version: {release_version}
        ```

        > **NOTE:** Replace only `develop/` so `develop/tests` becomes `tests/`.

        Every component image is published with a version defined in the `RELEASE_VERSION` file stored in the `test-infra` repository on the given release branch. Test scripts for integration jobs like GKE Integration or GKE Upgrade are also loaded from the `test-infra` release branch.

    2. Check all `yaml` files for the following references:  
        `image: eu.gcr.io/kyma-project/develop/{IMAGE_NAME}:{SOME_SHA}`
        Change the references to:  
        `image: eu.gcr.io/kyma-project/{IMAGE_NAME}:{release_version}`

        - `installation/resources/installer-local.yaml`

>         - In `installation/resources/installer.yaml` change from `eu.gcr.io/kyma-project/develop/installer:{image_tag}` to `eu.gcr.io/kyma-project/kyma-installer:{release_version}`

Reviewer: ```suggestion
        - In `installation/resources/installer.yaml` replace`eu.gcr.io/kyma-project/develop/installer:{image_tag}` with `eu.gcr.io/kyma-project/kyma-installer:{release_version}`
```

>     3. In `resources/core/values.yaml` file, replace the `clusterDocsTopicsVersion` value to your release branch name. For example, for 0.9.1 release, find the following section:

Reviewer: ```suggestion
    3. In  the `resources/core/values.yaml` file, replace the `clusterDocsTopicsVersion` value with your release branch name. For example, for the 0.9.1 release, find the following section:
```

        ```yaml
        docs:
            # (...) - truncated
            clusterDocsTopicsVersion: master
        ```

>         Replace the value `clusterDocsTopicsVersion` to the following:

Reviewer: ```suggestion
        And replace the `clusterDocsTopicsVersion` value with the following:
```

        ```yaml
        dir:
        version: {release_version}
        ```

        > **NOTE:** Replace only `develop/` so `develop/tests` becomes `tests/`.

        Every component image is published with a version defined in the `RELEASE_VERSION` file stored in the `test-infra` repository on the given release branch. Test scripts for integration jobs like GKE Integration or GKE Upgrade are also loaded from the `test-infra` release branch.

>     2. Check all `yaml` files for the following references:  

Reviewer: So actually you replace `image: eu.gcr.io/kyma-project/develop/{IMAGE_NAME}:{SOME_SHA}` with `image: eu.gcr.io/kyma-project/{IMAGE_NAME}:{release_version}` in `installation/resources/installer-local.yaml` and In `eu.gcr.io/kyma-project/develop/installer:{image_tag}` with `eu.gcr.io/kyma-project/kyma-installer:{release_version}` in `installation/resources/installer.yaml`? Maybe we make it clearer by making two seperate points of that (ii and iii)?

Author: What about now? 😄

>         `image: eu.gcr.io/kyma-project/develop/{IMAGE_NAME}:{SOME_SHA}`

Reviewer: Add some additional space here as the below renders in the same line

7. After all checks pass, merge the PR.
    > **NOTE:** To merge the PR to the release branch, you must receive approvals from all teams.

8. Merging the PR to the release branch runs the postsubmit job that creates a GitHub release. This job also triggers documentation update on the official Kyma website.

> 9. Update `RELEASE_VERSION` to the next patch RC1 version on the release branch. Do it immediately after the release, otherwise any PR to a release branch overrides the previously published Docker images.

Reviewer: I don't get the next patch RC1 version.

Author: yes, it always refer to a file.

Reviewer: so
```suggestion
9. Update the `RELEASE_VERSION` file to contain the next patch RC1 version on the release branch. Do it immediately after the release, otherwise, any PR to a release branch overrides the previously published Docker images.
```

>     For example, if `RELEASE_VERSION` on release branch contains `0.9.1`, change it to `0.9.2-rc1`.

Reviewer: ```suggestion
    For example, if the `RELEASE_VERSION` on the release branch is `0.9.1`, change it to `0.9.2-rc1`.
```

10.  Validate the `yaml` and changelog files generated under [releases](https://github.com/kyma-project/kyma/releases).

> 11. Update the release content manually with the links how to install the latest Kyma release.  

Reviewer: ```suggestion
11. Update the release content manually with links to the instruction on how to install the latest Kyma release.  
```

>     Currently this means to grab the links from the previous release and update the version number in URLs. If contributors want you to change something in them they would address you directly.

Reviewer: Do contributors want to change something in the links or in the instruction?
```suggestion
    Currently, this means to grab the links from the previous release and update the version number in URLs. If contributors want you to change something in the instruction, they would address you directly.
```
